### PR TITLE
Introduce RightArrow (->) token

### LINKED ISSUE #9605: Disallow spaces in the Yul return value arrow ``- >``.
I guess the fact that this is allowed is rather a parsing accident than actually intentional? I only noticed, since there's one semantics test ``test/libsolidity/semanticTests/inlineAssembly/inline_assembly_function_call.sol``, that uses this (I guess due to a typo).

### PR DESCRIPTION
**Pull request by axic**
Fixes #9605.

### CODE REVIEW

File: liblangutil/Token.h [hunk 1]
```diff
@@ -270,6 +270,7 @@ namespace solidity::langutil
 	\
 	/* Yul-specific tokens, but not keywords. */                       \
 	T(Leave, "leave", 0)                                               \
+	T(RightArrow, "->", 0)                                             \
```
Comment (axic) [inline liblangutil/Token.h]: We already have an `Arrow` token for `=>`. Should think about an alternative/better name, and no, not `YulArrow`.

File: liblangutil/Scanner.cpp [hunk 2]
```diff
@@ -563,12 +563,14 @@ void Scanner::scanToken()
...
 			break;
 		case '-':
-			// - -- -=
+			// - -- -= ->
 			advance();
 			if (m_char == '-')
 				token = selectToken(Token::Dec);
 			else if (m_char == '=')
 				token = selectToken(Token::AssignSub);
+			else if (m_char == '>' && m_kind == ScannerKind::Yul)
```
Comment (axic) [inline liblangutil/Scanner.cpp]: I actually think `->` is invalid in Solidity too, so could just enable this new token for both modes.

Comment (axic) [reply liblangutil/Scanner.cpp]: Ran all the tests with this enabled for Solidity, not a single failure.

File: liblangutil/Token.h [hunk 3]
```diff
@@ -83,7 +83,8 @@ namespace solidity::langutil
 	T(Semicolon, ";", 0)                                                \
 	T(Period, ".", 0)                                                   \
 	T(Conditional, "?", 3)                                              \
-	T(Arrow, "=>", 0)                                                   \
+	T(DoubleArrow, "=>", 0)                                             \
+	T(RightArrow, "->", 0)                                              \
```
Comment (axic) [inline liblangutil/Token.h:87]: Rust calls these FatArrow and RArrow (https://doc.rust-lang.org/reference/tokens.html#punctuation). Javascript does not have them.

File: docs/grammar/SolidityLexer.g4 [hunk 4]
```diff
@@ -111,7 +111,8 @@ Colon: ':';
 Semicolon: ';';
 Period: '.';
 Conditional: '?';
-Arrow: '=>';
+DoubleArrow: '=>';
+RightArrow: '->';
```
Comment (axic) [inline docs/grammar/SolidityLexer.g4:115]: This is here for the lexer to pick it up.

File: docs/grammar/SolidityLexer.g4 [hunk 5]
```diff
@@ -283,9 +284,7 @@ YulRParen: ')';
 YulAssign: ':=';
 YulPeriod: '.';
 YulComma: ',';
-// TODO: remove whitespace workaround once the parser disallows it.
-//@doc:name ->
-YulArrow: '->' | '-' YulWS+ '>';
+YulArrow: '->';
```
Comment (axic) [inline docs/grammar/SolidityLexer.g4:287]: Not sure if we can share tokens between the two lexer modes, hence kept it here with its own name.

File: Changelog.md [hunk 6]
```diff
@@ -15,6 +15,7 @@ Compiler Features:
 Bugfixes:
  * AST: Remove ``null`` member values also when the compiler is used in standard-json-mode.
  * Optimizer: Keep side-effects of ``x`` in ``byte(a, shr(b, x))`` even if the constants ``a`` and ``b`` would make the expression zero unconditionally. This optimizer rule is very hard if not impossible to trigger in a way that it can result in invalid code, though.
+ * Scanner: Fix bug where whitespace would be allowed within the ``->`` token (e.g. ``-   >`` becomes invalid now).
```
Comment (axic) [inline Changelog.md]: Maybe need to be more specific with the example: `e.g. function f() -  > x {}` becomes invalid in Yul.

**Comment by axic** [comment]
Done, I think... I hope.

**Review by leonardoalt — APPROVED** [review]
lgtm

File: docs/grammar/SolidityLexer.g4 (hunk 4 above)
Comment (chriseth) [reply docs/grammar/SolidityLexer.g4:115]: Does this have implications for the AST output? If Yes, I would prefer not changing `Arrow`, but it certainly has to be mentioned in its own changelog entry.

Comment (axic) [reply docs/grammar/SolidityLexer.g4:115]: It is an internal detail, isn't it? How would the AST be affected? None of the AST test emit a difference, but that could just mean tests are bad.

**Comment by chriseth** [comment]
Ok, I also don't see a place where this could be non-internal.

**Pull request merged**